packages/ckeditor5-engine/src/conversion/downcastdispatcher.js

```diff
@@ -489,6 +553,155 @@ export default class DowncastDispatcher {
+	 * Get changes without those that needs to be converted using {@link #convertRefresh} defined by a `triggerBy` configuration for
+	 * {@link module:engine/conversion/downcasthelpers~DowncastHelpers#elementToElement `elementToElement()`} conversion helper.
+	 *
+	 * @param {module:engine/model/differ~Differ} differ The differ object with buffered changes.
+	 * @returns {Array.<Object>}
+	 * @private
+	 */
+	_getChangesAfterAutomaticRefreshing( differ ) {
+		const changes = differ.getChanges();
+		const elementsToRefresh = this._getElementsForAutomaticRefresh( changes );
+
+		if ( !elementsToRefresh.size ) {
+			return changes;
+		}
+
+		for ( const element of elementsToRefresh.values() ) {
+			differ.refreshItem( element );
+		}
+
+		// The `differ.refreshItem()` invalidates differ cache - we can't re-use previous changes.
+		const changesAfterRefresh = differ.getChanges();
+
+		return changesAfterRefresh.filter( entry => !elementsToRefresh.has( getElementFromChange( entry ) ) );
+	}
+
+	/**
+	 * Returns elements that should be converted using {@link #convertRefresh} defined by a `triggerBy` configuration for
+	 * {@link module:engine/conversion/downcasthelpers~DowncastHelpers#elementToElement `elementToElement()`} conversion helper.
+	 *
+	 * This will return:
+	 *
+	 * * Element which attributes changed for an 'attribute' change.
+	 * * Parent of inserted or removed element for matched 'insert' or 'remove' changes.
+	 *
+	 * @param {Array.<Object>} changes The changes diff set from the differ.
+	 * @returns {Set.<module:engine/model/element~Element>}
+	 * @private
+	 */
+	_getElementsForAutomaticRefresh( changes ) {
+		const found = changes
+			.filter( ( { type } ) => type === 'attribute' || type === 'insert' || type === 'remove' )
+			.map( entry => {
+				const element = getElementFromChange( entry );
+
+				let eventName;
+
+				if ( entry.type === 'attribute' ) {
+					if ( !element ) {
+						// Refreshing is done only on elements so skip text attribute changes.
+						return;
+					}
+
+					eventName = `attribute:${ entry.attributeKey }:${ element.name }`;
+				} else {
+					eventName = `${ entry.type }:${ entry.name }`;
+				}
+
+				if ( this._isRefreshTriggerEvent( eventName, element.name ) ) {
+					return element;
+				}
+			} )
+			.filter( element => !!element );
+
+		return new Set( found );
+	}
+
+	/**
+	 * Handles reconverting a model element that has an existing model-to-view mapping.
+	 *
+	 * It performs a shallow conversion for the element and its attributes. All children that already have a converted view
+	 * will not be converted again. Their existing view elements will be used instead.
+	 *
+	 * @private
+	 * @param {Object} data Event data.
+	 */
+	_reconvertElement( data ) {
+		const currentView = this.conversionApi.mapper.toViewElement( data.item );
+
+		// Remove the old view but do not remove mapper mappings - those will be used to revive existing elements.
+		this.conversionApi.writer.remove( currentView );
```

Comment:
This is an optimistic assumption ;). This will work for most cases but may fail for more complex ones if model element is mapped to inner view element (something like in list items and \<ul>/\<ol> + \<li> I wonder if it works there correctly if I refresh a singular list item and also change its attribute :thinking: ).

I understand that the problem with "old refreshing" was that remove conversion removed mappings?

Reply:
Yes. Those are mappings that were removed. But I think that if would handle memoization on the upper level (ie by taking a snapshot before the whole conversion) or just by moving the wholel logic to dispatcher alone it would be better and we would need "refresh" change at all.the